cocotb/handle.py

```diff
@@ -414,40 +386,61 @@ def _setcachedvalue(self, *args, **kwargs):
     def __le__(self, *args, **kwargs):
         raise ValueError("Not permissible to set values on a constant object")
 
-class NonConstantObject(NonHierarchyObject):
+class NonHierarchyIndexableObject(NonHierarchyObject):
     def __init__(self, handle):
         """
             Args:
                 _handle [integer] : vpi/vhpi handle to the simulator object
         """
         NonHierarchyObject.__init__(self, handle)
-        self._r_edge = _RisingEdge(self)
-        self._f_edge = _FallingEdge(self)
-        self._e_edge = _Edge(self)
-
-    def __hash__(self):
-        return self._handle
+        self._range_left  = simulator.get_range_left(self._handle)
```

Comment:
@stuarthodgson and @chiggs 
I changed this interface slightly in a later version such that the simulator returns a tuple for the range or None if it is not indexable and there is no need for the indexable parameter.

@stuarthodgson I am not sure what you mean by handling the indexing translation in GPI.  The goal was to make indexing in Python look just like indexing in the HDL file, i.e. sig[23:16] should be indexed from python as sig[23] and not sig[7].  You are right that the index could just be passed and checked by the native_check_create(), but like @chiggs said, you need to know the range to generate the appropriate indices for iteration.